cordova-plugin-outline/apple/src/OutlinePlugin.swift

```diff
@@ -325,31 +325,42 @@ class OutlinePlugin: CDVPlugin {
     }
 
     DDLogInfo("Migrating UIWebView local storage to WKWebView")
-    var success = copy(from: uiWebViewLocalStorage.relativePath, to: wkWebViewLocalStorage.relativePath)
-    DDLogInfo("\t\(uiWebViewLocalStorage) => \(wkWebViewLocalStorage) => \(success)")
-    success = copy(from: "\(uiWebViewLocalStorage.relativePath)-shm", to: "\(wkWebViewLocalStorage.relativePath)-shm")
-    DDLogDebug("\tSHM => \(success)")
-    success = copy(from: "\(uiWebViewLocalStorage.relativePath)-wal", to: "\(wkWebViewLocalStorage.relativePath)-wal")
-    DDLogDebug("\tWAL => \(success)")
+    if !copyAtomic(from: uiWebViewLocalStorage, to: wkWebViewLocalStorage) {
+      DDLogError("Failed to migrate local storage: \(uiWebViewLocalStorage) => \(wkWebViewLocalStorage)")
+      return
+    }
+    if !copyAtomic(from: URL.init(fileURLWithPath: "\(uiWebViewLocalStorage.relativePath)-shm"),
+                   to: URL.init(fileURLWithPath: "\(wkWebViewLocalStorage.relativePath)-shm")) {
+      DDLogWarn("Failed to migrate .localstorage-shm file")
+    }
+    if !copyAtomic(from: URL.init(fileURLWithPath: "\(uiWebViewLocalStorage.relativePath)-wal"),
+                   to: URL.init(fileURLWithPath: "\(wkWebViewLocalStorage.relativePath)-wal")) {
+      DDLogWarn("Failed to migrate .localstorage-wal file")
+    }
   }
 
-  // Copies a file from |src| to |dest|. Returns whether the operation was successful.
-  func copy(from src: String, to dest: String) -> Bool {
+  // Atomically copies a file from |src| to |dest|. Returns whether the operation was successful.
+  func copyAtomic(from src: URL, to dest: URL) -> Bool {
     let fileManager = FileManager.default
     // Source file does not exist
-    if !fileManager.fileExists(atPath: src) {
+    if !fileManager.fileExists(atPath: src.relativePath) {
+      DDLogError("Source file does not exist: \(src)")
       return false
     }
     // Destination file exists
-    if fileManager.fileExists(atPath: dest) {
+    if fileManager.fileExists(atPath: dest.relativePath) {
+      DDLogError("Destination file exists: \(dest)")
       return false
     }
-    // Create destination path and copy src to dst
-    let destDir = URL(fileURLWithPath: dest).deletingLastPathComponent().relativePath
+
+    // Create destination path and atomically write the contents of |src| to |dst|
+    let destDir = dest.deletingLastPathComponent()
     do {
-      try fileManager.createDirectory(atPath: destDir, withIntermediateDirectories: true, attributes: nil)
-      try fileManager.copyItem(atPath: src, toPath: dest)
+      let data = try Data.init(contentsOf: src)
```

Comment:
My understanding is that this will read the entire content of the file in memory, then write it atomically. Now you need to have memory available to keep the entire file in memory, which wasn't the case before, since the library probably writes as it reads, consuming memory independent of the file size. How big is that file?

In any case, the migration is still not atomic. If one of the files fail, or the app crashes, you have a corrupted state.

What I had in mind was:
1. Copy all the files to a temporary directory. Don't use the atomic write to keep memory independent of the file size.
1. Once all files are copied to the temp dir successfully, move the dir to the new location. The directory move is usually atomic.

If there are other files in `LocalStorage/` you need to keep, you would have to copy those too. Perhaps you can start by copying `LocalStorage/`, and modifying the files there, then moving back.

Reply:
>> How big is that file?

The .localstorage file is a sqlite3 file that, for our app, should be a few KB at most.


>> In any case, the migration is still not atomic. If one of the files fail, or the app crashes, you have a corrupted state.

The -wal and -shm files are [temp files](https://sqlite.org/tempfiles.html) used by sqlite. They are recreated on DB access. Failing to copy them is not an issue. Whether the migration succeeds depends on the atomic copying of `file__0.localstorage`. 

I think this approach is fine, but I'll implement your suggestion.

Reply:
Implemented your suggestion in 4a432e9. Turns out I was wrong about the auxiliary files (-wal, -shm) - without them, the app is not able to read the main local storage file. I believe the migration is now atomic.